// ngx-print-element/esm2022/lib/ngx-print-element.directive.mjs
import { Directive } from '@angular/core';
import * as i0 from "@angular/core";
import * as i1 from "./ngx-print-element.service";
export class NgxPrintElementDirective {
    renderer;
    prints;
    constructor(renderer, prints) {
        this.renderer = renderer;
        this.prints = prints;
    }
    ngOnInit() {
        this.prints.renderer = this.renderer;
    }
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.2.1", ngImport: i0, type: NgxPrintElementDirective, deps: [{ token: i0.Renderer2 }, { token: i1.NgxPrintElementService }], target: i0.ɵɵFactoryTarget.Directive });
    static ɵdir = i0.ɵɵngDeclareDirective({ minVersion: "14.0.0", version: "18.2.1", type: NgxPrintElementDirective, isStandalone: true, selector: "[ngxPrintElement]", ngImport: i0 });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.2.1", ngImport: i0, type: NgxPrintElementDirective, decorators: [{
            type: Directive,
            args: [{
                    standalone: true,
                    selector: '[ngxPrintElement]'
                }]
        }], ctorParameters: () => [{ type: i0.Renderer2 }, { type: i1.NgxPrintElementService }] });
