// ngx-print-element/esm2022/lib/ngx-print-element.component.mjs
import { Component } from '@angular/core';
import * as i0 from "@angular/core";
import * as i1 from "./ngx-print-element.service";
export class NgxPrintElementComponent {
    prints;
    constructor(prints) {
        this.prints = prints;
    }
    /**
     * Print
     * @param elementRef
     * @param config
     * @returns
     */
    print(elementRef, config) {
        return this.prints.print(elementRef, config);
    }
    static ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "18.2.1", ngImport: i0, type: NgxPrintElementComponent, deps: [{ token: i1.NgxPrintElementService }], target: i0.ɵɵFactoryTarget.Component });
    static ɵcmp = i0.ɵɵngDeclareComponent({ minVersion: "14.0.0", version: "18.2.1", type: NgxPrintElementComponent, isStandalone: true, selector: "ngx-print-element", exportAs: ["element"], ngImport: i0, template: `<ng-content></ng-content>`, isInline: true });
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "18.2.1", ngImport: i0, type: NgxPrintElementComponent, decorators: [{
            type: Component,
            args: [{
                    selector: 'ngx-print-element',
                    standalone: true,
                    imports: [],
                    template: `<ng-content></ng-content>`,
                    exportAs: 'element'
                }]
        }], ctorParameters: () => [{ type: i1.NgxPrintElementService }] });
